go/apps/newfileproxy/common/services.go

```diff
@@ -0,0 +1,927 @@
+		return
+	}
+	err = info.DeleteOwner(*combinedPrincipal, &serverData.ResourceMutex)
+	if err == nil {
+		SuccessResponse(ms, ServiceType_DELETEOWNER)
+	} else {
+		FailureResponse(ms, ServiceType_DELETEOWNER, "Can't insert resource")
+	}
+	return
+}
+
+func DoDeleteReader(ms *util.MessageStream, serverData *ServerData, connectionData *ServerConnectionData,
+		msg FileproxyMessage) {
+	if len(msg.Arguments) < 1 {
+		FailureResponse(ms, ServiceType_DELETEREADER, "Not enough arguments")
+		return
+	}
+	resourceName := msg.Arguments[0]
+	info := serverData.ResourceManager.FindResource(resourceName, &serverData.ResourceMutex)
+	if info == nil {
+		FailureResponse(ms, ServiceType_DELETEREADER, "no such resource")
+		return
+	}
+	if !IsAuthorized(*msg.TypeOfService, serverData, connectionData, info) {
+		FailureResponse(ms, ServiceType_DELETEREADER, "not authorized")
+		return
+	}
+	combinedPrincipal, err :=  GetCombinedPrincipal(msg.Data)
+	if err != nil {
+		FailureResponse(ms, ServiceType_DELETEREADER, "Can't parse combined principal")
+		return
+	}
+	err = info.DeleteWriter(*combinedPrincipal, &serverData.ResourceMutex)
+	if err == nil {
+		SuccessResponse(ms, ServiceType_DELETEREADER)
+	} else {
+		FailureResponse(ms, ServiceType_DELETEREADER, "Can't delete")
+	}
+	return
+}
+
+func DoDeleteWriter(ms *util.MessageStream, serverData *ServerData, connectionData *ServerConnectionData,
+		msg FileproxyMessage) {
+	if len(msg.Arguments) < 1 {
+		FailureResponse(ms, ServiceType_DELETEWRITER, "Not enough arguments")
+		return
+	}
+	resourceName := msg.Arguments[0]
+	info := serverData.ResourceManager.FindResource(resourceName, &serverData.ResourceMutex)
+	if info == nil {
+		FailureResponse(ms, ServiceType_DELETEWRITER, "no such resource")
+		return
+	}
+	if !IsAuthorized(*msg.TypeOfService, serverData, connectionData, info) {
+		FailureResponse(ms, ServiceType_DELETEWRITER, "not authorized")
+		return
+	}
+	combinedPrincipal, err :=  GetCombinedPrincipal(msg.Data)
+	if err != nil {
+		FailureResponse(ms, ServiceType_DELETEWRITER, "Can't parse combined principal")
+		return
+	}
+	err = info.DeleteWriter(*combinedPrincipal, &serverData.ResourceMutex)
+	if err == nil {
+		SuccessResponse(ms, ServiceType_DELETEWRITER)
+	} else {
+		FailureResponse(ms, ServiceType_DELETEWRITER, "Can't delete resource")
+	}
+	return
+}
+
+func DoReadResource(ms *util.MessageStream, serverData *ServerData, connectionData *ServerConnectionData,
+		msg FileproxyMessage) {
+	if len(msg.Arguments) < 1 {
+		FailureResponse(ms, ServiceType_READ, "Not enough arguments")
+		return
+	}
+	resourceName := msg.Arguments[0]
+	info := serverData.ResourceManager.FindResource(resourceName, &serverData.ResourceMutex)
+	if info == nil {
+		FailureResponse(ms, ServiceType_READ, "no such resource")
+		return
+	}
+	if !IsAuthorized(*msg.TypeOfService, serverData, connectionData, info) {
+		FailureResponse(ms, ServiceType_READ, "not authorized")
+		return
+	}
+	SendFile(ms, serverData, info)
```

Comment:
I thought SendFile should return an error here, but I guess it will do FailureResponse internally.